File: ta/ref-app/android/ClickCall/src/com/wx2/clickcall/Calabash.java
package com.wx2.clickcall;

import com.webex.wme.MediaConnection;
import com.webex.wme.MediaTrack;

import java.util.ArrayList;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Created by dan on 8/14/14.
 */
public class Calabash {
    private ArrayList<OnMediaReadyArgs> onMediaReadyInfoList = new ArrayList<OnMediaReadyArgs>();
    private JSONObject mCSIChangeHistory = new JSONObject();

    public static class OnMediaReadyArgs {
        int mid;
        MediaConnection.MediaDirection dir;
        MediaConnection.MediaType type;
        MediaTrack track;
    }
    
    public String GetCSIChangeHistory() {
    	return mCSIChangeHistory.toString();
    }

    /**
     * onMediaReady: Store and verify callback variables
     * @param mid
     * @param dir
     * @param type
     * @param track
     * @return
     */
    public void onMediaReadyStore (int mid, MediaConnection.MediaDirection dir, MediaConnection.MediaType type, MediaTrack track) {
        OnMediaReadyArgs myArgs = new OnMediaReadyArgs();
        myArgs.mid = mid;
        myArgs.dir = dir;
        myArgs.type = type;
        myArgs.track = track;

        onMediaReadyInfoList.add(myArgs);
    }

    public void onCSIChanged(String mediaType, long vid, long[] newCSIArray) {
		try {
	    	if(!mCSIChangeHistory.has(mediaType))
					mCSIChangeHistory.put(mediaType, new JSONObject());
	    	JSONObject csiMedia = mCSIChangeHistory.optJSONObject(mediaType);
	    	String sVid = String.valueOf(vid);
	    	if(!csiMedia.has(sVid))
	    		csiMedia.put(sVid, new JSONArray());
	    	JSONArray csiArray = csiMedia.optJSONArray(sVid);
	    	JSONArray newCSIs = new JSONArray();
	    	for(long csi : newCSIArray) {
	    		newCSIs.put(csi);
	    	}
	    	csiArray.put(newCSIs);
		} catch (JSONException e) {
			e.printStackTrace();
		}
    }

    public ArrayList<OnMediaReadyArgs> getOnMediaReadyInfoList() {
        return onMediaReadyInfoList;
    }
}

File: ta/ref-app/src/Calabash.cpp
//
//  Calabash.cpp
//  MediaSessionTest
//
//  Created by Dan Lee Jo on 8/15/14.
//  Copyright (c) 2014 Dan Lee Jo. All rights reserved.
//

#include "Calabash.h"

Calabash::Calabash() : m_CSIChangedHistory(Json::objectValue)
{
    m_CSIChangedHistory["audio"] = Json::Value(Json::objectValue);
    m_CSIChangedHistory["video"] = Json::Value(Json::objectValue);
    m_CSIChangedHistory["share"] = Json::Value(Json::objectValue);
}

Calabash::~Calabash() {

}

void Calabash::onMediaReadyStore (unsigned long mid, WmeDirection direction, WmeSessionType mediaType, IWmeMediaTrackBase *pTrack) {
    OnMediaReadyArgs myArgs;
    myArgs.mid = mid;
    myArgs.direction = direction;
    myArgs.mediaType = mediaType;
    myArgs.pTrack = pTrack;

    onMediaReadyInfoList.push_back(myArgs);

}

std::string Calabash::GetCSIChangeHistory()
{
    Json::FastWriter w;
    return w.write(m_CSIChangedHistory);
}

void Calabash::onCSIChanged(const char* mediaType, unsigned int vid, const unsigned int* newCSIArray, unsigned int newCSICount)
{
    std::string sVid = std::to_string(vid);
    Json::Value & csi_media = m_CSIChangedHistory[mediaType];
    if (!csi_media.isMember(sVid))
        csi_media[sVid] = Json::Value(Json::arrayValue);
    Json::Value & csi_array = csi_media[sVid];
    Json::Value newCSIs(Json::arrayValue);
    int i = 0;
    for (i = 0; i < newCSICount; i++) {
        newCSIs.append(Json::Value(newCSIArray[i]));
    }
    csi_array.append(newCSIs);
}

std::vector<Calabash::OnMediaReadyArgs> Calabash::getOnMediaReadyInfoList() {
    return onMediaReadyInfoList;
}
